### DISCUSSION TITLE
Preview Feedback: C# 12 Alias any type

### DISCUSSION THREAD

**Discussion by KathleenDollard** [dotnet/csharplang#7110] [General]
Aliases for any type (except nullable reference types) are included starting in C# 12 preview 3. 

You can use this discussion for feedback as you use the feature.

You can learn more about aliases for any type in the article [What's new in C# 12](https://learn.microsoft.com/dotnet/csharp/whats-new/csharp-12#alias-any-type) article.

**Comment by fforjan** [comment]
Brainstorming here, I didn't have time to try but looking for feedback. We have an issue where our applications is case insensitive, meaning when we compare names, we cannot use the == operator. We have done a roslyn analyzer rule which check for the Name property of a specific interface. Any other property is not checked.
Could we use alias in that case like using OurObjectIdentifierType = string; and then update our roslyn analyzer rule to check for the == usage on a OurObjectIdentifierType  type on any property ? 
In order word, what is the integration with rolsyn analyzer ?

**Reply by 333fred** [reply]
Roslyn analyzers have access to the full symbol model, including seeing things like `IAliasSymbol`s. Nothing about this feature affects what they can see or would affect your using aliases for that scenario. A better place to have this conversation is probably on https://discord.gg/csharp, in the roslyn channel, where you can talk to multiple developers on the roslyn project about how you might accomplish this.

**Reply by fforjan** [reply]
Thanks for the info! I didn't dknow about the discord !

**Comment by BartoszChrostowski** [comment]
[Porposal](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/using-alias-types) suggests the change:
> The order in which using_alias_directives are written has no significance, and resolution of the `(namespace_name | type)` referenced by a using_alias_directive is not affected by the using_alias_directive itself or by other using_directives in the immediately containing compilation unit or namespace body. In other words, the `(namespace_name | type)` of a using_alias_directive is resolved as if the immediately containing compilation unit or namespace body had no using_directives. A using_alias_directive may however be affected by extern_alias_directives in the immediately containing compilation unit or namespace body.

which highly limits the use case for such an aliasing system to the readability within one compilation unit. However, if we were to take two separate compilation units, it seems like it would make it harder to understand. Force developers to read using directives to understand why we can put ("m", 100) in place of Measurement (as in example `using Measurement = (string Units, int Distance)`)

Are there any plans to further extend the types system? It would be reasonable to allow such aliasing that would allow type naming to propagate outside the compilation unit. It would allow users to define such things as already for example `string = System.String` is themselves globally and would solve the problem I have described above.

**Reply by CyrusNajmabadi** [reply]
> [Porposal](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/using-alias-types) suggests the change:

This has been the logic for using aliases since their inception.  The proposal doesn't change anything wrt to that.

> It would be reasonable to allow such aliasing that would allow type naming to propagate outside the compilation unit.

You can allow aliases to propagate outside a compilation unit by making them `global` :)

**Reply by BartoszChrostowski** [reply]
What I am trying to say is that I am afraid that the fact it does not propagate may lead to some serious mess.
Given two compilation units:

> using Measurement = (string Units, int Distance)
> public int Foo(Measurement measurement) => Boo(measurement);

> using Distance = (string DistanceUnit, int Distance)
> public int Boo(Distance distance)

You will have to understand what Measurement and Distance are. 

What differs between aliases for namespaces and types is that in the case of namespaces, the feature allows using types with the same name from different namespaces in a readable way. The name of the type does not change. You can more or less guess what's the underlying type (the only mistake you can make is guessing the wrong namespace). In the case of type aliases, you can do stuff that may make it impossible.

**Comment by omariom** [comment]
> You can allow aliases to propagate outside a compilation unit by making them global

What!? 😮
How did I miss that?

**Reply by BartoszChrostowski** [reply]
[C#10 what's new article, global usings](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10#global-using-directives)

**Reply by JiriZidek** [reply]
Do I understand well that this is still a syntactic suger with no impact on assembly DLL ? So there is no need for public/private logic, since after compilation there are no "type aliases" like Measurement ??

**Reply by CyrusNajmabadi** [reply]
This is exactly the same as current aliases, just that what you can alias has been expanded to any type, not just named-types.

**Comment by keithn** [comment]
Problem with an alias is it isn't a strong type, so from https://devblogs.microsoft.com/dotnet/check-out-csharp-12-preview/?WT.mc_id=DOP-MVP-5001655 where they show ... using DatabaseInt = int?;   

DatabaseInt isn't protected such that you couldn't assign it  QueueInt.

**Reply by CyrusNajmabadi** [reply]
@keithn that's unrelated to this feature, and is an aspect of aliases in general.  If you want strong typing, wrap the underlying type with a struct-record.  Aliases are just aliases.  They're another name to reference a type with, but which is otherwise exactly the same as that type.

**Reply by keithn** [reply]
I think it actually is, wrapping the underlying type is kind of awful and horrible way to go.  Thing is, if you did something like a typedef of things like ints in the future, it would pretty much look exactly like alias's, in fact people are already using alias like a typedef except it has no type checking...  and I get you guys are knee deep in this whole alias business to take a step back, but I think you are creating a mess with alias.

**Reply by keithn** [reply]
in fact read the article announcing these features  "Using directives for additional types"  .... it's trying to say to use it as a way of creating type definitions that aren't really types....  the examples are things that should be type checked....   I know this is going to fall on deaf ears because the team is so firmly held in the "It's just an alias!!!"  and won't recognize that its main use is going to be a typedef and is already advocated as essentially a typedef.     The article says to give feedbac, this is it.  Alias's are a poor mans typedef and really you should have proper typedefs instead.

**Reply by CyrusNajmabadi** [reply]
> I think it actually is

It isn't.  The aspect of aliases you are discussing was part of their design back when they were added in *C# 2* (which was *18* years ago).

> and I get you guys are knee deep in this whole alias business to take a step back, but I think you are creating a mess with alias.

You are talking about a part of it's design from nearly two decades ago.  This discussion is about the simple relaxation we did so that when you do `using X = ...` the thing on in the `...` portion doesn't have arbitrary limitations like it did back in C#2.  This change is strictly a relaxation, allowing for more use cases, not changing how the existing alias would work.  Thanks :)

**Reply by keithn** [reply]
again, I point you to the release blog and see how it is being advocated as a poor mans typedef.  I get to you its just about alias's, but what it is creating is a mess.... that's my feedback.

**Reply by CyrusNajmabadi** [reply]
Def green free to give that feedback to that blog post. The feature here isn't about typedefs.  It's just about alias relaxations. :-)

Note, we likely have proposals for things closer to typedefs. Def add your feedback there. Thanks!

**Reply by BartoszChrostowski** [reply]
I believe I [argue the same](https://github.com/dotnet/csharplang/discussions/7110#discussioncomment-5585337).

An I agree that the stuff from example should be strongly typed and should not be used for presentation purposes. This may lead for misuses. The only thing i can think off, where it may have some potential is for example in mappings. Instead of 
> using Dto = Foo.Api
> using Domain = Foo.Domain

and use `Dto.YourBooClass` and `Domain.YourBooClass`

You may write
> using YourBooClassDto = Foo.Api.YourBooClass
> using YourBooClassDomain = Foo.Domain.YourBooClass

But idk why would you prefer the latter one

**Reply by tannergooding** [reply]
The ability for someone to misuse a feature is not really a good argument for why something shouldn't exist. If that were true, we wouldn't have programming at all because people also "misuse" ints, strings, lists, etc.

They can equally simplify code, make it more readable, and support powerful forms of feature enablement.

Consider that you can do something like this to support either `double` or `float` without having to maintain 2 copies of the code. Using the newer generic math feature with `where T : IBinaryFloatingPointIeee754<T>` is another alternative.
```csharp
#if FEATURE_DOUBLE
using float_t = double;
#else
using float_t = float;
#endif
```

Consider that you may simply have some form of tuple used privately that you want to have a nice temporary name for:
```csharp
using NameAgePair = (string Name, int Age);
```

Consider that you may have a longer generic where writing it out every time is troublesome:
```csharp
using CostAPISignature = System.Func<Player, FixedPoint, LoadedActionConfig, bool>;
using PropertySet = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<string>>;
```

And the list goes on. Just as there are many examples where you can "hurt" things with aliases, there are just as many where selective usage can greatly improve things and we actively use them in such places in the BCL as we have millions of lines of code that need to be actively maintained, kept performant, and kept generally readable at the same time.

**Reply by BartoszChrostowski** [reply]
```cs
#if FEATURE_DOUBLE
using float_t = double;
#else
using float_t = float;
#endif
```

You'll have to put it as `global`, but I see some point. My problem with this feature is that it will be only file-wise, or put as `global`, which has its own problem. It would be accessed no matter the namespace you are in.

Moreover, I would argue
```cs
using NameAgePair = (string Name, int Age);
```
Should be `record struct`

And 
```cs
using CostAPISignature = System.Func<Player, FixedPoint, LoadedActionConfig, bool>;
```
should be `delegate`

But yeah, I also acknowledge it would be easier to read aliases instead of the long generic name. But it would be even easier if it wasn't constrained to the compilation unit it is defined in.

**Reply by CyrusNajmabadi** [reply]
> And
> using CostAPISignature = System.Func<Player, FixedPoint, LoadedActionConfig, bool>;

Delegates are nominal types.  This would force people away from structural typing for Func/Action, which would push us backwards against the very thing we worked towards 15+ years ago due to all the problems around nominal typing here.  

This is the benefit of aliases.  They can give you good names, without going into the *nominal typing* space.

**Reply by CyrusNajmabadi** [reply]
> Should be record struct

Why?  Now you have another nominal type that is incompatible with simple tuples.  Not every time one want a good name does that mean they want a *full type* :)  

Aliases exist precisely because, since C# 2, there has long been a space for just names, without new types.  Nothing in that regard changes here.  Just a simple relaxation around restrictions that really didn't have any particular justification behind them, and whcih teh community gave *lots* of feedback on over hte years that htey wanted removed.

**Reply by tannergooding** [reply]
> Now you have another nominal type that is incompatible with simple tuples

This is also undesirable when you're simply trying to use a tuple to improve some internal implementation details and where a full type is just too much.

Not everything is public surface area and not everything needs to have the full surface implemented.

**Reply by CyrusNajmabadi** [reply]
@tannergooding agreed.

**Reply by BartoszChrostowski** [reply]
>Delegates are nominal types[...]

My bad. The delegate part was stupid. I'm sorry.
>Not every time one wants a good name does that mean they want a full type

It is hard for me to imagine an example in which I would prefer tuple. Can you provide some example?
>Aliases exist precisely because, since C# 2[...]

I mean, that is a good argument in favour of this feature that this is just a relaxation. I am probably not fond of those aliases overall 😄

Thank you for your explanations and discussion.

**Comment by olivermue** [comment]
I'm also seeing, that this feature will be highly misused or misinterpret by most casual programmers. Instead of a simple alias that could lead to misinterpretation, I would favor real derived types (maybe with constraints) like in F#. By that way we could create types like `ValidatedString` that will be created by functions like `ValidatedString Validate(string maybeEvilMessage)` and ensure all business logic methods only consume a `ValidatedString` instead of `string`.

By creating the alias type feature people are going to create exactly this kind of code and feeling safe, but it will be only an alias and still allows to call `Process(ValidatedString message)` with a simple `string` leading to undesired behaviour. So please really consider to drop this feature. What would really help would be to create easily alias types like `public class ValidatedString : string` to use the type system also on a business layer to differenciate between the different usages of string, int, Guid, etc.

**Reply by JustinMinnaar** [reply]
This should be possible to enable or disable this feature in the solution or project file, as it will cause problems. I would agree that a strictly typed alias / definition would be safer and would also be useful.

**Reply by En3Tho** [reply]
F#'s aliases are exactly like showcased C# aliases.
```fsharp
type ThisIsJustAnAlias = int
```
is the same as
```csharp
using ThisIsJustAndAlias = int
```
What you might refer to is "Single case DU" which are basically record structs with F#-specific syntax and have nothing common with aliases.

**Reply by HaloFour** [reply]
C# has always had aliases and it's always been possible to misuse them.  C# even defines some of it's own, in that `string` is really `System.String` and you can use those names interchangeably.  This proposal doesn't change anything about the nature of aliases, it only allows them to be used in a few scenarios combined with other language features that have not yet supported aliases, such as named tuples.

Type-safe aliases would be a nice feature to have, and there are several proposals for it.  You can use a simple wrapper struct record today, which covers much of the boilerplate, and it's possible that "explicit extensions" might be an even more powerful way to achieve it in the future.

**Reply by BartoszChrostowski** [reply]
> even defines some of it's own, in that string is really System.String

Yes. But it is defined on a language level (in its specification). It would be nicer if such aliases would not be constrained to compilation unit. Thus it would be somewhat possible to remove the `string` definition from c# language specification and define `data string = System.String` at `global` namespace or something. (_I know that this is simplification_).

**Reply by CyrusNajmabadi** [reply]
> and define data string = System.String at global namespace or something.

This wouldn't work as that would then allow others to provide their own alias of `string` within their namespace that would take precedence over the global one.  This is something we'd actively be against.  It's a design *goal* of the built-in keywords that you know *100%* of the time in absolutely all contexts exactly what `string` means :)

**Reply by BartoszChrostowski** [reply]
Yeah yeah, as said; simplification. I understand that it is not feasible

**Reply by CosminSontu** [reply]
A> > and define data string = System.String at global namespace or something.
> 
> This wouldn't work as that would then allow others to provide their own alias of `string` within their namespace that would take precedence over the global one. This is something we'd actively be against. It's a design _goal_ of the built-in keywords that you know _100%_ of the time in absolutely all contexts exactly what `string` means :)

You can prevent special type names from being used as aliases, instead of having a limited type aliasing feature in the language.
I would like to define behaviour (e. g. Via extension methods) on the alias only, and not on the aliased type.


This was mentioned before...
If I alias String with EmailAddress, 
String GetDomain(this EmailAddress ea)... method, should be callable only on EmailAddress and never on String. This is not the case now...
Also global usings requires the clients of your library to issue those global usings in their project, otherwise aliases won't work.

The question is... Will there be another language feature / concept that will address these points?

**Reply by CyrusNajmabadi** [reply]
> The question is... Will there be another language feature / concept that will address these points?

Def open a discussion of you're interested in that :-)

**Reply by CosminSontu** [reply]
> > The question is... Will there be another language feature / concept that will address these points?
> 
> Def open a discussion of you're interested in that :-)

The language team should be interested in that, don't you think?

**Reply by HaloFour** [reply]
> The language team should be interested in that, don't you think?

They are, which is why there is an active working group on the topic:  https://github.com/dotnet/csharplang/tree/main/meetings/working-groups/roles

**Reply by CosminSontu** [reply]
> > The language team should be interested in that, don't you think?
> 
> They are, which is why there is an active working group on the topic: https://github.com/dotnet/csharplang/tree/main/meetings/working-groups/roles

interesting and complicated proposal... instead of going towards roles and views, why not properly separate type aliasing with a typedef construct and let extension methods do their thing?
one would need to control creation of alias types hence the need to specify constructors / factory methods for these types.
extension methods defined on a type alias must not be invoke-able on types being aliased. This would help with type-driven development mindset.
e.g.
 typedef EmailAddress = String
.. 
string GetDomain(this EmailAddress @this) => ...
...
EmailAddress ea = EmailAddress.Create("abc@def.gh"); //calling private constructor here
string mail = "abc@def.gh";
var domain = ea.GetDomain(); // works
var domain1 = mail.GetDomain(); //compiler error

**Reply by CyrusNajmabadi** [reply]
> why not properly separate type aliasing with a typedef construct and let extension methods do their thing?

That would limit things just to extension methods.  It would also mean you wouldn't get any functionality just referencing the type-def, you'd also have to be sure to pull in the extensions.

**Reply by CosminSontu** [reply]
> > why not properly separate type aliasing with a typedef construct and let extension methods do their thing?
> 
> That would limit things just to extension methods. It would also mean you wouldn't get any functionality just referencing the type-def, you'd also have to be sure to pull in the extensions.

that's correct, but this is just a limited example of what could be.
ideally you could define extra methods on type aliases.
ideally you could define the extension methods directly inside the namespaces
so they are imported together with type aliases... 
a better propagation of aliases and extension methods than current
_using static_ and _global using_
I'm not thinking about limiting anything here :)

**Reply by CyrusNajmabadi** [reply]
> ideally you could define extra methods on type aliases.

Now we're just defining roles/extensions.  :)

> so they are imported together with type aliases...

That's what an extension would be.  Just without having to figure out some sort of 'imported together' approach.  :)

> a better propagation of aliases and extension methods than current

This still doesn't work as you can't extend things like type parameters then.  And you don't get properties, indexers, and other members, etc. Extensions are a way to bridge all of this, with a syntax that is very in line with normal .net types :)

**Reply by HaloFour** [reply]
The roles/extensions proposal enables that, plus also enables other types of members, such as extension properties, and extension implementations of other interfaces.  Theoretically they could also function as active patterns.

**Reply by Richiban** [reply]
@HaloFour 

> Type-safe aliases would be a nice feature to have, and there are several proposals for it. You can use a simple wrapper struct record today

Unfortunately records don't help with this use case, because you can't write constructor validation (at least not without manually overriding writing both the constructor and value property, which is half the reason to use a record).

I know it's off topic for this thread but something I've wanted in C# for _so long_ is simply this:

```cs
public record EmailAddress(string Value)
{
    {
        if (!Regex.Match(Value, "[^@]+@[^@]\.[^@]+"))
            throw new ArgumentException("Not a valid email");
    }
}
```

This pattern is vital for domain driven design and I still have to write it out manually, which makes me has a sad :(.

**Reply by Joe4evr** [reply]
@Richiban I believe the "validator"/"final initializer" feature to make that possible is still on the language team's radar, just delayed again.

**Reply by HaloFour** [reply]
@Richiban 

> Unfortunately records don't help with this use case, because you can't write constructor validation (at least not without manually overriding writing both the constructor and value property, which is half the reason to use a record).

Single field structs work just as well, for the same reason (the runtime doesn't know the difference).  Sure, it's more code than a simple record, but you only have to write it once.  Or you could have a source generator do the heavy lifting delegating the validation to a partial method or the like.

I'm hoping that "explicit extensions" get something like active patterns that would force validation before you can treat an instance of the underlying type as the extension.

**Reply by CosminSontu** [reply]
> > > The question is... Will there be another language feature / concept that will address these points?
> > 
> > 
> > Def open a discussion of you're interested in that :-)
> 
> The language team should be interested in that, don't you think?

And the language team is really listening :)
I was watching Microsoft Build 2023 : What’s new in C# 12 and beyond | BRK203H presentation on yt (https://youtu.be/hq6uEaMYf5U?t=2366) and at minute 39 Mads and Dustin showcase this code sample:

```
public explicit extension JsonString for string 
{
       public DataObject Parse() { ... parse as Json ... }
}

void Display(JsonString json) { ... json.Parse() ... }

Display("""{ "customer" : { "name" : "Microsoft" ... } }""");
```
This is pretty much what I was describing when I said "letting extension methods do their thing". This is the right approach since it does not add complexity. The "for" there is akin to how Rust impl blocks are defined. I think this is great!

**Reply by JiriZidek** [reply]
But without typedefs allowing things like "Tee" | "Cofee", or ranges (0...100) it is just half of the problem. These "types" should have overridable static methods (aka convertors from their primitive parent) to do gate guard of values into the "type" (by throwing OutOfRange).

**Reply by Richiban** [reply]
@HaloFour 

Looking back on it now, I kind of wish the team hadn't implemented records as a language feature; I would much rather they had implemented primary constructors (as they are currently, in preview) and then some out-of-the-box source generators for equality, public properties etc (and maybe a `Record` super generator that combined them all). That would be a much more flexible and extensible state to be in now, I feel.

**Reply by HaloFour** [reply]
@Richiban 

Possibly.  I brought that up and members of the team felt that the use cases were compelling enough to deserve their own first-class language feature.  But the fact that records are a language feature doesn't preclude using source generators to design your own that behave as you would expect.  Although you couldn't use `with` as that feature is still tightly tied to records and hasn't yet been decomposed into a separate feature.

**Reply by CyrusNajmabadi** [reply]
>  That would be a much more flexible and extensible state to be in now, I feel.

That is still possible today.  Almost every part of a record is something the user can customize.  So you can def still use generators for your flexibility/extensibility use cases.  This was def something we considered and felt was a good way to handle this situation.

**Comment by JiriZidek** [comment]
I only sorry that the keyword is not "type" like in majority of languages.

**Reply by 333fred** [reply]
Which keyword do you mean?

**Reply by BreyerW** [reply]
@333fred almost certainly he means alias not being treated like type (so you cant assign string to aliased string and vice versa).

 And indeed it would be very useful for domain driven projects. I really hope next step for some future c# version will be to make this a possibility

Using record wrappers is unreliable for performance critical code: regularly in runtime are found bugs with promoting single field structs. Situation improves from version to version but i wouldnt recommend this approach to anyone who either dont have time to dissect asm code or cant yet

**Reply by En3Tho** [reply]
No I think he literally meant that keyword is "using" instead of "type"
For example F# or TypeScript: ```type Xxx = Yyy```

**Reply by CyrusNajmabadi** [reply]
This is the syntax C# has had since C# 2.0 (Back in 2005).  :)

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

> This is the syntax C# has had since C# 2.0 (Back in 2005). :)

It's been in the language since the beginning, it's in the [ECMA-334 1st edition (December 2001)](https://www.ecma-international.org/wp-content/uploads/ECMA-334_1st_edition_december_2001.pdf), 16.3.1 Using alias directives.

**Reply by CyrusNajmabadi** [reply]
Ah, you're right.  I was thinking of another feature we did in 2.0 (namespace aliases `::`, and got the timelines confused).

**Reply by JiriZidek** [reply]
> Which keyword do you mean?

"using"

But by further reading I realized, that the feature is probably not about defining new "types" (like in Turbo Pascal or TypeScript), but rather just macro definition. If that is the case, I'd reconsider this, and "using" looks appropriate.

**Reply by CosminSontu** [reply]
> > Which keyword do you mean?
> 
> "using"
> 
> But by further reading I realized, that the feature is probably not about defining new "types" (like in Turbo Pascal or TypeScript), but rather just macro definition. If that is the case, I'd reconsider this, and "using" looks appropriate.

Indeed, a typedef would be needed, specially for library authoring. Global usings have to be placed in all client projects of the library to work as intended.
A typedef would create an alias type and this alias would be imported with the library... Also, you could define behavour against the alias and the behaviour would be applicable only to the alias. 
This could be a nice entrypoint for union types... 
typedef EmailAddress = String
typedef OptionalPerson = Person | Nobody

just dreaming...

**Comment by AlwaysHC** [comment]
Instead of just "alias" I'd like to see type checking.

**Comment by huoyaoyuan** [comment]
Please note that this is NOT a brand new language feature. It just fills the gap for the already existing feature since C# 2.
Aliasing to create separated types was discussed for a long time, and is likely to be covered by roles/extensions.

**Reply by keithn** [reply]
sure.... I'm not sure you are breaking out of your bubble enough to see the problem though.  It is already being promoted, by microsoft, as way to create pseudo types, the examples given are what should really be typedefs, they are the examples given, you are enabling a mess.  But you guys only see "This is just a feature we already have and we are opening it up a bit more"  .... opening it up a bit more to create a mess of types that aren't types.   I get it, you don't want to listen to that and think it's all fine, but the very article, by microsoft, promoting pseduo/messy types is the one tell us to come here to provide feedback.   If you just want to defend the feature against the feedback, that's your choice, but I think its clear from the feedback that really you should do something different to get typedefs before you do this so we don't end up with a mess.     Telling us you discussed it a long time ago and done nothing about it just makes it worse :)  You do see that opening it up lets people, because it is the only way to do it, create types from things like primitives that aren't type checked?  You do see that is how it is being promoted already?

**Reply by CyrusNajmabadi** [reply]
> he feedback that really you should do something different to get typedefs before you do this so we don't end up with a mess. 

Aliases *already* exist and shipped.  This happened back in C# 1.0.  This language change is just simplifying a restriction the language had that wasn't really necessary.  It's no more and no less than that :)

> Telling us you discussed it a long time ago and done nothing about it just makes it worse

We haven't done 'nothing' about it.  THere is full scale work going on in this space *right now* both in design *and* in terms of prototype implementations.  I'm not sure how you got that idea that because we did a little simplication here that we're not working on features like the ones you're asking for :)

> You do see that opening it up lets people, because it is the only way to do it, create types from things like primitives that aren't type checked?

This was *always* possible in C#.  You could always create non-checked type aliases to primitive types.  You could do that in C# 1.0.  Nothing about this relaxation changes that.  :)

**Reply by En3Tho** [reply]
It's clear from your feedback that you want entirely different feature without any spec attached.

How do those strict aliases work? How do they interoperate with base type, with each other? How do they work with instance methods of base type, how do they work with static methods that take base type? How do you cast base type to this new one? How do you validate it is actually a correct value?

There are lots of questions and they should belong to a different discussion altogether because it's a totally different feature that resembles roles (a new feature) and has nothing to do with C#'s aliases. Calling it typedefs or any other term from another language doesn't make sense because it is not that.

**Comment by Bellarmine-Head** [comment]
From this distance, I can't see myself using the `alias any type` feature that much (could be wrong; you never know!), but I'm very grateful to this discussion for putting me on to the `newtype` form:-

```csharp
public readonly record struct AuthToken(String Token);
```

That's something that I've wanted in C# for a long time, but I had thus far overlooked this feature of C# 10.

**Comment by HaufenKlaus** [comment]
Wouldn't a private struct record be the same as an alias? I don't get it, C# Language features were about having to write less, now it is about writing things differently. I'm still waiting to for actually useful semi-auto properties that have been promised since C# 10 (?) How about you stop botching easy to implement bloatware on top and do the work that is actually useful... You were a bit slow before C# 8 when it comes to implementing features, now you can't differ quantity from quality....

**Reply by CyrusNajmabadi** [reply]
> Wouldn't a private struct record be the same as an alias? 

No.  That would be a type with a different name.  An alias is fully erased.  It isn't anything new.  It's jsut a name you can use in code as a simpler/different way to refer to a type or namespace (but is still that type or namespace).

> Wouldn't a private struct record be the same as an alias? 

Aliases let you write less.

> now it is about writing things differently.

Aliases have existed since C# 1.0.  This was just relaxing a rule we had in the language that the RHS had certain annoying limits on it (limits that became more annoying as the language has added more and more types).

> I'm still waiting to for actually useful semi-auto properties that has been promised since C# 10

That is still be working on.  

> and do the stuff that is actually useful...

This feature was one of hte most requested and highest voted ones in all the language.  What may not be useful to you is still useful to the rest of the community.  And, as i mentioned, we continue to work on 'semi auto properties'.  I'm the author of that specification and it's something i can say we've been working hard on.

**Reply by HaufenKlaus** [reply]
OK, I look into that, thanks for that info ;-)

**Reply by CyrusNajmabadi** [reply]
You're welcome :)

**Comment by vladimir-kovalyuk** [comment]
I'm looking for the feature that would help addressing the Primitive Obsession problem. There are a lot of expositions of this problem in different areas. For instance in terms of identity of the objects we used to identify them via Guid or integral type. The method that accepts identities of multiple different entities opens the possibility for the errors where entities are mixed, for instance:
```
void Foo(Guid departmentId, Guid employeeId);
Foo(employee.Id, department.Id);
```
Or units of measurement suffer with the same problem:
```
double miles = 100;
double kilometers = 10;
miles += kilometers;

```
Unfortunately aliases do not help with primitive obsession. Shall I create another proposal?

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/discussions/6406
https://github.com/dotnet/csharplang/discussions/6681
https://github.com/dotnet/csharplang/discussions/7031

**Reply by JiriZidek** [reply]
F# solves

**Comment by carnitron** [comment]
I have one request I'd like to throw in related to using aliases for any type.  Maybe it's a separate feature, I'm not sure.  I haven't and won't have time in the immediate future to put together a super tight spec, so forgive me if I have overlooked something.

Request:
I would like to be able put using statements in any scope, not just at the top of the file.  And the aliases would apply to just that scope.

Reasons:

1. I want to make an alias to a nested type in a nested class, but since I have to put it at the top of the file, I have to expose it outside of the parent class, if it's even possible in the situation.

2. I often use nested static classes to fully qualify big banks of constants and similar static functions.  Fully qualifying them ensures clarity and prevents naming collisions.  But it is a lot to type.  I know this is why you added "using static" syntax.  But all of those usings are up at the top of the file.  So you lose the clarity they bring at the point of usage.  If you could put them NEXT to the point of usage, you'd get the best of both worlds: minimal typing AND clarity.

3. When I use the "using static" syntax to minimize what I have to type, it's easy to get into situations where can't tell what constant/function/etc. is backed by an alias and where it came from.  Since the fully qualified name is gone, the syntax often makes it look like the thing I am referring to is local to the class and not from somewhere else.

4. Sometimes naming collisions in like namespaces prevent you from being able to remove all of your naming qualifications.


Example:

So this:

```
public void Thing(double[] numbers)
{
    double average = A.B.C.D.E.F.Metric.Average(numbers);
    double stdDev = A.B.C.D.E.F.Metric.StandardDeviation(numbers);
    // ... Many more metrics.
}
```
Or this:

```
// Way at the top of the file.
using static A.B.C.D.E.F.Metric;

// Pages of code later:
public void Thing(double[] numbers)
{
    double average = Average(numbers);
    double stdDev = StandardDeviation(numbers);
}
```
Would become this:

```
public void Thing(double[] numbers)
{
    using static A.B.C.D.E.F;

    double average = Metric.Average(numbers);
    double stdDev = Metric.StandardDeviation(numbers);
}
```
Or even better, this:

```
public void Thing(double[] numbers)
{
    // Adding the "connecting" definition here makes it completely unambiguous what functions I'm using at the
    // call site.
    using static Metric = A.B.C.D.E.F.Metric;

    double average = Metric.Average(numbers);
    double stdDev = Metric.StandardDeviation(numbers);
}
```
Hope this all makes sense.  Thx for reading.

**Reply by CyrusNajmabadi** [reply]
Hi @carnitron please open a new discussion on that idea. Thanks!

**Reply by carnitron** [reply]
Sure!

**Reply by carnitron** [reply]
Here:
https://github.com/dotnet/csharplang/discussions/7238

**Comment by carnitron** [comment]
FWIW, I like this proposal.

I am constantly waging a war to reduce visual clutter and eliminate complex expressions and overly long lines.  People talk about being explicit to increase clarity, but the positives quickly turn into negatives when you start combining tuples, generic types and fully qualified names.  Visually parsing the super explicit thing you are looking at becomes hard and sometimes even requires temporary on-the-spot editing to break statements apart to simply read it all, let alone understand what is happening.

I love tuples but don't use them as much as I'd hoped in some cases b/c of this issue.  Hopefully this proposal nails that.

I do think a type system would be good too.  C# pretty much systematically improves and simplifies nearly everything in C++ and it is almost comical how much easier the language is to use and how much more empowering it is; I probably write C# 2x-5x faster than I write C++ code, with far fewer bugs.

The one area where this hasn't been true for me so far is usings vs typedefs.  (Though usings are obviously way better than includes.)  I have found myself trying all sorts of things to work around some of the issues here, and ftm have settled on writing generic classes to have little subclass wrappers for different type combinations to give them simpler names.

It sounds like you guys are working on something long term there, which would be awesome. 

Personally, I love that the C# team continually makes improvements to the language that make me more productive and reduce what *I* have to write.  Complexity is a tricky thing; but I'd rather have the complexity be in learning lots of simple features than entangled in a large codebase that would be simplified by those new features.  Less is more for MY code, not the language feature set (within reason, ofc).

So keep it up.  It is absolutely understandable that some of the features the community wants are complicated and have to be thought through and implemented carefully, perhaps over multiple versions. And in the mean time, it is still good to deliver value via easier, simpler features.

Cheers.

**Comment by Shadowblitz16** [comment]
Can we get a way to alias types inside namespaces?

I would like to wrap up some types for my game engine instead of having to reference the library or write my own

**Comment by AntonGavrilov** [comment]
Hello, can we make alias for static methods? So we can add global using for Console methods for example: Console.WriteLine method (global using WriteLine = System.Console.WriteLine;) and after this we can use just WriteLine everywhere insted of Console.WriteLine.

**Reply by CyrusNajmabadi** [reply]
You can do that with `global using static System.Console;`

**Comment by TylerBrinkley** [comment]
I know the C# 12 ship has sailed but I was sad to see that this alias any type feature doesn't support generic types. For instance, where I work we have a different notion of what a `Task` is. I've explored getting us to rename it due to the conflict but it's baked in pretty heavily and is a central part of our system, so we've defined a global using alias for `System.Threading.Tasks.Task` as `SysTask`. Unfortunately, we cannot do the same for `System.Threading.Tasks.Task<T>` for consistency and so we use `Task<T>` and `SysTask` throughout our codebase which is a bit unfortunate but not a huge deal.

**Reply by CyrusNajmabadi** [reply]
@TylerBrinkley that is tracked already with https://github.com/dotnet/csharplang/issues/1239

**Reply by TylerBrinkley** [reply]
Thank you

**Comment by RenderMichael** [comment]
Has there been discussion on `using static <type keyword>` such as `using static string`? As a minimal strawman:

```csharp
using static string;

var combined = Join(' ', 1, 2, 3);
```

Today we have to do `using static System.String`.

This would be well in line with aspects of C# 12's `alias any type` such as aliasing tuples instead of having to use `ValueTuple` directly.

**Reply by 333fred** [reply]
I would champion such a proposal.

**Reply by Perksey** [reply]
This would be cool. Would it just be keywords, or would anything that could be accepted as an argument to `typeof` be accepted as well?

**Comment by CyrusNajmabadi** [comment]
Note: this was not done because the cases where static members arise are basically identical to the cases where named type can be used.  So there was no 'gap' per se.  saying, for example, `using static System.String;` works equally as well, with the exact same semantics.

**Comment by Hermholtz** [comment]
FYI, I wanted to alias the following type:
```
using ValueReader = System.Func<byte[], ref int, float>;
```

but it doesn't accept `ref` (CS1073: unexpected token 'ref'). The following is syntactically okay:

```
using ValueReader = System.Func<byte[], int, float>;
```

but for my scenario `ref` would be helpful.

I also tried `out` (just to check if it works) and it's also not accepted, albeit with a different error message:

```
using ValueReader = System.Func<byte[], out int, float>;
```
> CS1960: Invalid variance modifier. Only interface and delegate type parameters can be specified as variant.

Maybe in the future it could be improved to accept `ref` and `out`. Thanks.

**Reply by HaloFour** [reply]
That's not related to aliasing.  You can't change the "ref"-ness of a type through generics.  You'd need a different generic delegate that explicitly defined the parameter as `ref`.

**Comment by Hermholtz** [comment]
So, is it "alias any type" feature, or "alias any type, but"?

**Reply by sharpchen** [reply]
`Func` and `Action` don't support `ref`, `out`, `params` and other modifiers, it's not an issue about type alias

**Reply by Hermholtz** [reply]
Thank you, that explanation I understand. I wasn't aware about this.

**Reply by HaloFour** [reply]
You could define a delegate with that specific signature:

```cs
public delegate float ValueReader(byte[] arg1, out int arg2);
```

Then you can use that for lambdas, callbacks, etc.

You could also define it as generic, but you'd have to make the second param `ref`/`out`:

```cs
public delegate TR OutFunc<T1, T2, TR>(T1 arg1, out T2 arg2);
```

and then alias that:

```cs
using ValueReader = OutFunc<byte[], int, float>;
```
